Update CLOUD_PROVIDER for iptables installation

ajaypvictor opened the pull request:
Fixes https://github.com/confidential-containers/cloud-api-adaptor/issues/2304

Extending the CLOUD_PROVIDER check for installing iptables which fixes the APF failure seen with libvirt provider.
Also, wrapping `setup-nat-for-imds` service only for `azure, aws and generic` providers.

stevenhorsman commented on src/cloud-api-adaptor/podvm/qcow2/misc-settings.sh:
```diff
@@ -44,8 +44,8 @@ if [[ "$CLOUD_PROVIDER" == "azure" || "$CLOUD_PROVIDER" == "generic" ]] && [[ "$
-# Setup oneshot systemd service for AWS and Azure to enable NAT rules
-if [ "$CLOUD_PROVIDER" == "azure" ] || [ "$CLOUD_PROVIDER" == "aws" ] || [ "$CLOUD_PROVIDER" == "generic" ]
+# Install iptables on all providers except docker/vsphere.
+if [ "$CLOUD_PROVIDER" != "vsphere" ] || [ "$CLOUD_PROVIDER" != "docker" ]
```
I think this needs to be an and here?
```suggestion
if [ "$CLOUD_PROVIDER" != "vsphere" ] && [ "$CLOUD_PROVIDER" != "docker" ]
```
This might also have been me giving poor guidance as an example. I think this rule is okay, but it would be good to hear from others

ajaypvictor replied:
oops, apologies idea was to give `==` rather than `!=`, let me get it fixed.. thanks for the suggestion.

ajaypvictor commented:
Tested with a rhel qcow2 image on `s390x`, but apparently it's not having any yum repos configured to install `iptables`:
```
==> qemu.rhel: Uploading qcow2/misc-settings.sh => ~/misc-settings.sh
==> qemu.rhel: Provisioning with shell script: /tmp/packer-shell3726067260
    qemu.rhel: Starting the RHEL flow.
==> qemu.rhel: Error: There are no enabled repositories in "/etc/yum.repos.d", "/etc/yum/repos.d", "/etc/distro.repos.d".
==> qemu.rhel: Removed "/etc/systemd/system/multi-user.target.wants/kdump.service".
==> qemu.rhel: Removed "/etc/systemd/system/multi-user.target.wants/tuned.service".
==> qemu.rhel: Failed to disable unit: Unit file firewalld.service does not exist.
==> qemu.rhel: Uploading qcow2/build-s390x-se-image.sh => ~/build-s390x-se-image.sh
==> qemu.rhel: Provisioning with shell script: /tmp/packer-shell2706290340
    qemu.rhel: Building SE podvm image for s390x
```

stevenhorsman reviewed (approved):
LGTM. Thanks @ajaypvictor!

ajaypvictor commented:
Test results with the PR changes:
```
    qemu.rhel: The yum/dnf plugins: /etc/dnf/plugins/subscription-manager.conf, /etc/dnf/plugins/product-id.conf were automatically enabled for the benefit of Red Hat Subscription Management. If not desired, use "subscription-manager config --rhsm.auto_enable_yum_plugins=0" to block this behavior.
    qemu.rhel:
    qemu.rhel: Updating Subscription Management repositories.
    qemu.rhel: Waiting for process with pid 4324 to finish.

    qemu.rhel: Red Hat Enterprise Linux 9 for IBM z Systems -  2.6 MB/s |  47 MB     00:17
    qemu.rhel: Red Hat Enterprise Linux 9 for IBM z Systems -  2.5 MB/s |  21 MB     00:08
    qemu.rhel: Last metadata expiration check: 0:00:05 ago on Wed Feb 26 08:59:50 2025.
    qemu.rhel: Dependencies resolved.
    qemu.rhel: ================================================================================
    qemu.rhel:  Package         Arch    Version           Repository                      Size
    qemu.rhel: ================================================================================
    qemu.rhel: Installing:
    qemu.rhel:  iptables-nft    s390x   1.8.10-11.el9_5   rhel-9-for-s390x-baseos-rpms   210 k
    qemu.rhel: Upgrading:
    qemu.rhel:  iptables-libs   s390x   1.8.10-11.el9_5   rhel-9-for-s390x-baseos-rpms   456 k
    qemu.rhel: Installing dependencies:
    qemu.rhel:  libnftnl        s390x   1.2.6-4.el9_4     rhel-9-for-s390x-baseos-rpms    84 k
    qemu.rhel:
    qemu.rhel: Transaction Summary
    qemu.rhel: ================================================================================
    qemu.rhel: Install  2 Packages
    qemu.rhel: Upgrade  1 Package
    qemu.rhel:
    qemu.rhel: Total download size: 750 k
    qemu.rhel: Downloading Packages:
    qemu.rhel: (1/3): libnftnl-1.2.6-4.el9_4.s390x.rpm          93 kB/s |  84 kB     00:00
    qemu.rhel: (2/3): iptables-nft-1.8.10-11.el9_5.s390x.rpm   194 kB/s | 210 kB     00:01
    qemu.rhel: (3/3): iptables-libs-1.8.10-11.el9_5.s390x.rpm  319 kB/s | 456 kB     00:01
    qemu.rhel: --------------------------------------------------------------------------------
    qemu.rhel: Total                                           502 kB/s | 750 kB     00:01
    qemu.rhel: Red Hat Enterprise Linux 9 for IBM z Systems -  199 kB/s | 3.6 kB     00:00
==> qemu.rhel: Importing GPG key 0xFD431D51:
==> qemu.rhel:  Userid     : "Red Hat, Inc. (release key 2) <security@redhat.com>"
==> qemu.rhel:  Fingerprint: 567E 347A D004 4ADE 55BA 8A5F 199E 2F91 FD43 1D51
==> qemu.rhel:  From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release
    qemu.rhel: Key imported successfully
==> qemu.rhel: Importing GPG key 0x5A6340B3:
==> qemu.rhel:  Userid     : "Red Hat, Inc. (auxiliary key 3) <security@redhat.com>"
==> qemu.rhel:  Fingerprint: 7E46 2425 8C40 6535 D56D 6F13 5054 E4A4 5A63 40B3
==> qemu.rhel:  From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-redhat-release
    qemu.rhel: Key imported successfully
    qemu.rhel: Running transaction check
    qemu.rhel: Transaction check succeeded.
    qemu.rhel: Running transaction test
    qemu.rhel: Transaction test succeeded.
    qemu.rhel: Running transaction
    qemu.rhel:   Preparing        :                                                        1/1
    qemu.rhel:   Upgrading        : iptables-libs-1.8.10-11.el9_5.s390x                    1/4
    qemu.rhel:   Installing       : libnftnl-1.2.6-4.el9_4.s390x                           2/4
    qemu.rhel:   Installing       : iptables-nft-1.8.10-11.el9_5.s390x                     3/4
    qemu.rhel:   Running scriptlet: iptables-nft-1.8.10-11.el9_5.s390x                     3/4
    qemu.rhel:   Cleanup          : iptables-libs-1.8.10-4.el9_4.s390x                     4/4
    qemu.rhel:   Running scriptlet: iptables-libs-1.8.10-4.el9_4.s390x                     4/4
    qemu.rhel:   Verifying        : libnftnl-1.2.6-4.el9_4.s390x                           1/4
    qemu.rhel:   Verifying        : iptables-nft-1.8.10-11.el9_5.s390x                     2/4
    qemu.rhel:   Verifying        : iptables-libs-1.8.10-11.el9_5.s390x                    3/4
    qemu.rhel:   Verifying        : iptables-libs-1.8.10-4.el9_4.s390x                     4/4
    qemu.rhel: Installed products updated.
    qemu.rhel:
    qemu.rhel: Upgraded:
    qemu.rhel:   iptables-libs-1.8.10-11.el9_5.s390x
    qemu.rhel: Installed:
    qemu.rhel:   iptables-nft-1.8.10-11.el9_5.s390x        libnftnl-1.2.6-4.el9_4.s390x
    qemu.rhel:
    qemu.rhel: Complete!
    qemu.rhel: Unregistering from: subscription.rhsm.redhat.com:443/subscription
    qemu.rhel: System has been unregistered.
==> qemu.rhel: Removed "/etc/systemd/system/multi-user.target.wants/kdump.service".
==> qemu.rhel: Removed "/etc/systemd/system/multi-user.target.wants/tuned.service".
==> qemu.rhel: Failed to disable unit: Unit file firewalld.service does not exist.
```

snir911 commented on src/cloud-api-adaptor/podvm/qcow2/misc-settings.sh:
```diff
@@ -44,13 +44,22 @@ if [[ "$CLOUD_PROVIDER" == "azure" || "$CLOUD_PROVIDER" == "generic" ]] && [[ "$
+                dnf -q install iptables -y; \
+                subscription-manager unregister
+            else
+                dnf -q install iptables -y;
```
iptables installation in rhel requires subscription if it's for s390 but not for other archs? have you tested this?
in any case, would it make the whole script running with subscription when key is provided (instead of only for iptables) ? and unregister in the end

ajaypvictor replied:
Apologies, i have not tested on x86 setup, let me try testing it on x86..
as of now only for the `iptables` (and it's dependencies) subscription is made, and then unregistered immediately once these packages are installed, do you see a scope of keeping the subscription for the whole script?

ajaypvictor replied:
Incorporated these, removed the check for s390x as well.

bpradipt commented on src/cloud-api-adaptor/podvm/qcow2/misc-settings.sh:
```diff
@@ -44,8 +49,8 @@ if [[ "$CLOUD_PROVIDER" == "azure" || "$CLOUD_PROVIDER" == "generic" ]] && [[ "$
 
-# Setup oneshot systemd service for AWS and Azure to enable NAT rules
-if [ "$CLOUD_PROVIDER" == "azure" ] || [ "$CLOUD_PROVIDER" == "aws" ] || [ "$CLOUD_PROVIDER" == "generic" ]
+# Install iptables for all providers except docker/vsphere.
```
Triggered a CI for docker to ensure this change doesn't break docker provider

ldoktor commented on src/cloud-api-adaptor/podvm/qcow2/misc-settings.sh:
```diff
@@ -15,6 +15,11 @@ echo -n | sudo tee /etc/machine-id
 
+# Subscribe RHEL incase of ACTIVATION_KEY & ORG_ID provided.
+if [[ -n "${ACTIVATION_KEY}" && -n "${ORG_ID}" ]]; then \
+    subscription-manager register --org=${ORG_ID} --activationkey=${ACTIVATION_KEY}
```
could you please quote the variables?

ldoktor reviewed (approved):
Thanks, it shouldn't change the behaviour unless key is provided. I haven't functionaly tested with the key but looks alright.

snir911 reviewed (approved):
Thanks!
